corehq/apps/dump_reload/management/commands/dump_domain_data_raw.py

```diff
@@ -0,0 +1,119 @@
+from __future__ import absolute_import, print_function
+
+import gzip
+import json
+import os
+import tempfile
+
+import boto3
+from datetime import datetime
+
+from botocore.exceptions import ClientError
+from django.conf import settings
+from django.core.management.base import BaseCommand, CommandError
+from s3transfer import S3Transfer
+
+from corehq.apps.es import FormES, CaseES
+from corehq.blobs.s3db import is_not_found
+from corehq.elastic import ES_EXPORT_INSTANCE
+from corehq.util.log import with_progress_bar
+
+
+class Command(BaseCommand):
+    """Created for the data mining project: https://github.com/dimagi/dimagi_defaulters
+    """
+    help = "Dump a raw JSON data to disk and upload to S3."
+
+    def add_arguments(self, parser):
+        parser.add_argument('domain')
+        parser.add_argument('-r', '--s3-region', dest='region', default='us-east-1')
+        parser.add_argument('-k', '--s3-key', dest='key')
+        parser.add_argument('-s', '--s3-secret', dest='secret')
+        parser.add_argument('-b', '--s3-bucket', dest='bucket')
+        parser.add_argument('-t', '--type', choices=['forms', 'cases'], dest='type')
+
+    def handle(self, domain, **options):
+        self.domain = domain
+        self.timestamp = datetime.utcnow()
+
+        if options['key']:
+            s3_key = options['key']
+            s3_secret = options['secret']
+        else:
+            s3_key = settings.S3_ACCESS_KEY
+            s3_secret = settings.S3_SECRET_KEY
+
+        if not (s3_key and s3_secret):
+            raise CommandError("S3 credentials not provided.")
+
+        self.bucket = options['bucket'] or 'raw-data-{}'.format(domain)
+        self.client = boto3.client(
+            's3', region_name=options['region'],
+            aws_access_key_id=s3_key, aws_secret_access_key=s3_secret
+        )
+        self._create_bucket()
+
+        exporters = {
+            'forms': _get_form_query,
+            'cases': _get_case_query
+        }.items()
+
+        if options['type']:
+            exporters = [exp for exp in exporters if exp[0] == options['type']]
+
+        for type_, query_fn in exporters:
+            query = query_fn(domain)
+            path = _dump_docs(query, type_)
+            self._upload(type_, path)
+            os.remove(path)
+
+    def _create_bucket(self):
```

Comment:
nit: could call this `_create_bucket_if_needed` or `_ensure_bucket_exists` or something.